stats/cloud/data.go

```diff
@@ -0,0 +1,134 @@
+}
+
+// NewSampleFromTrail just creates a ready-to-send Sample instance
+// directly from a netext.Trail.
+func NewSampleFromTrail(trail *netext.Trail) *Sample {
+	return &Sample{
+		Type:   "Points",
+		Metric: "http_req_li_all",
+		Data: SampleDataMap{
+			Time: trail.GetTime(),
+			Tags: trail.GetTags(),
+			Values: map[string]float64{
+				metrics.HTTPReqs.Name:        1,
+				metrics.HTTPReqDuration.Name: stats.D(trail.Duration),
+
+				metrics.HTTPReqBlocked.Name:        stats.D(trail.Blocked),
+				metrics.HTTPReqConnecting.Name:     stats.D(trail.Connecting),
+				metrics.HTTPReqTLSHandshaking.Name: stats.D(trail.TLSHandshaking),
+				metrics.HTTPReqSending.Name:        stats.D(trail.Sending),
+				metrics.HTTPReqWaiting.Name:        stats.D(trail.Waiting),
+				metrics.HTTPReqReceiving.Name:      stats.D(trail.Receiving),
+			},
+		},
+	}
+}
+
+// SampleDataAggregatedHTTPReqs is used in aggregated samples for HTTP requests.
+type SampleDataAggregatedHTTPReqs struct {
+	Time   time.Time         `json:"time"`
+	Type   string            `json:"type"`
+	Count  uint64            `json:"count"`
+	Tags   *stats.SampleTags `json:"tags,omitempty"`
+	Values struct {
+		Duration       AggregatedMetric `json:"http_req_duration"`
+		Blocked        AggregatedMetric `json:"http_req_blocked"`
+		Connecting     AggregatedMetric `json:"http_req_connecting"`
+		TLSHandshaking AggregatedMetric `json:"http_req_tls_handshaking"`
+		Sending        AggregatedMetric `json:"http_req_sending"`
+		Waiting        AggregatedMetric `json:"http_req_waiting"`
+		Receiving      AggregatedMetric `json:"http_req_receiving"`
+	} `json:"values"`
+}
+
+// CalcAverages calculates and sets all `Avg` properties in the `Values` struct
+func (sdagg *SampleDataAggregatedHTTPReqs) CalcAverages() {
+	count := float64(sdagg.Count)
+	sdagg.Values.Duration.Avg = float64(sdagg.Values.Duration.sum) / count
+	sdagg.Values.Blocked.Avg = float64(sdagg.Values.Blocked.sum) / count
+	sdagg.Values.Connecting.Avg = float64(sdagg.Values.Connecting.sum) / count
+	sdagg.Values.TLSHandshaking.Avg = float64(sdagg.Values.TLSHandshaking.sum) / count
+	sdagg.Values.Sending.Avg = float64(sdagg.Values.Sending.sum) / count
+	sdagg.Values.Waiting.Avg = float64(sdagg.Values.Waiting.sum) / count
+	sdagg.Values.Receiving.Avg = float64(sdagg.Values.Receiving.sum) / count
+}
+
+// AggregatedMetric is used to store aggregated information for a
+// particular metric in an SampleDataAggregatedMap.
+type AggregatedMetric struct {
+	Min time.Duration `json:"min"`
+	Max time.Duration `json:"max"`
+	sum time.Duration `json:"-"`   // ignored in JSON output because of SampleDataAggregatedHTTPReqs.Count
+	Avg float64       `json:"avg"` // not updated automatically, has to be set externally
+}
+
+// Add the new duration to the internal sum and update Min and Max if necessary
+func (am *AggregatedMetric) Add(t time.Duration) {
+	if am.sum == 0 || am.Min > t {
+		am.Min = t
```

Comment:
I think `am.sum == 0` is redundant condition.

Reply:
All variables in Go have a default value, and the default value for `time.Duration` (which [is basically `int64`](https://golang.org/src/time/time.go?s=21095:21114#L610)) is `0`. If I don't include some sort of other condition besides `am.Min > t`, `Min` will always be `0`

Reply:
👍